vcstools/gfit.py

```diff
@@ -664,148 +710,164 @@ def error_in_x_pos(pcov, popt, x):
         W10_e = np.sqrt(err_10_1**2 + err_10_2**2)
 
         # W50 root errors
-        err_50_1 = error_in_x_pos(pcov, popt, W50_roots[0])
-        err_50_2 = error_in_x_pos(pcov, popt, W50_roots[-1])
+        err_50_1 = error_in_x_pos(
+            x, self._fit_profile, self._noise_std, round(W50_roots[0]))
+        err_50_2 = error_in_x_pos(
+            x, self._fit_profile, self._noise_std, round(W50_roots[-1]))
         W50_e = np.sqrt(err_50_1**2 + err_50_2**2)
 
         # Wscat root errors
-        err_scat_1 = error_in_x_pos(pcov, popt, Wscat_roots[0])
-        err_scat_2 = error_in_x_pos(pcov, popt, Wscat_roots[-1])
+        err_scat_1 = error_in_x_pos(
+            x, self._fit_profile, self._noise_std, round(Wscat_roots[0]))
+        err_scat_2 = error_in_x_pos(
+            x, self._fit_profile, self._noise_std, round(Wscat_roots[-1]))
         Wscat_e = np.sqrt(err_scat_1**2 + err_scat_2**2)
 
         # Weq errors - using covariance formula
-        on_pulse_less = (on_pulse - std).clip(min=0)
-        spline0 = UnivariateSpline(x, on_pulse_less, s=0)
+        on_pulse_less = (on_pulse - self._noise_std).clip(min=0)
+        spline0 = UnivariateSpline(x_eq, on_pulse_less, s=0)
         integral = spline0.integral(0, len(self._std_profile)-1)
         dwdint = 1/max(on_pulse)**2
         dwdmax = -integral/max(on_pulse)**2
-        int_e = abs(integral/max(on_pulse - std) - integral/max(on_pulse))
-        max_e = std
-        Weq_e = np.sqrt( dwdint**2 * int_e**2 + dwdmax**2 * max_e**2 + 2*dwdint*dwdmax*int_e*max_e )
-
-        return [W10, W50, Weq, Wscat, W10_e, W50_e, Weq_e, Wscat_e]
+        int_e = abs(integral/max(on_pulse - self._noise_std) -
+                    integral/max(on_pulse))
+        max_e = self._noise_std
+        Weq_e = np.sqrt(dwdint**2 * int_e**2 + dwdmax**2 *
+                        max_e**2 + 2*dwdint*dwdmax*int_e*max_e)
+
+        # Convert to phases
+        self._W10 = W10/len(self._fit_profile)
+        self._W10_e = W10_e/len(self._fit_profile)
+        self._W50 = W50/len(self._fit_profile)
+        self._W50_e = W50_e/len(self._fit_profile)
+        self._Wscat = Wscat/len(self._fit_profile)
+        self._Wscat_e = Wscat_e/len(self._fit_profile)
+        self._Weq = Weq/len(self._fit_profile)
+        self._Weq_e = Weq_e/len(self._fit_profile)
+
+    def _est_noise_sn(self):
+        """
+        Estimates the SN of a profile using the fit profile to determine an off-pulse region.
+        """
+        sorted_profile = [i for _, i in sorted(
+            zip(self._fit_profile, self._std_profile), key=lambda pair: pair[0])]
+        ten_percent = round(0.1*len(self._std_profile))
+        self._noise_std = np.std(sorted_profile[:ten_percent])
+        self._noise_mean = np.mean(sorted_profile[:ten_percent])
+        self._sn = 1/self._noise_std
+        # TODO: make this estimate better
+        self._sn_e = 1/(self._noise_std * np.sqrt(2 * ten_percent - 2))
+
```

Comment:
Another thing I need to talk to you about is making a "flux SN" estimate. Basically, instead of the maximum signal divided by the noise, the "flux SN" is the area of the on pulse divided by the width (which is equivalent to the height) divided by the noise. This is what is used for flux density calculations as it more accurately the energy/area/flux of the profile.